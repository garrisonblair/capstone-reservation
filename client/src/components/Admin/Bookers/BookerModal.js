import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Modal, Dropdown, Button, List, Icon,
} from 'semantic-ui-react';
import sweetAlert from 'sweetalert2';
import api from '../../../utils/api';
import './BookerModal.scss';

class BookerModal extends Component {
  state = {
    myPrivileges: [],
    dropdownOptions: [],
    privilegeValue: '',
    privileges: [],
  }

  componentDidMount() {
    const { booker } = this.props;
    const { dropdownOptions } = this.state;
    this.setState({ myPrivileges: booker.privilege_categories });
    api.getPrivileges()
      .then((result) => {
        this.setState({ privileges: result.data });
        result.data.map(
          // eslint-disable-next-line array-callback-return
          (p) => {
            dropdownOptions.push({ text: p.name, value: p.id, key: p.id });
            this.setState({ dropdownOptions });
          },
        );
      });
  }

  handleDropdownChange = (e, { value }) => {
    this.setState({ privilegeValue: value });
  }

  handleRemovePrivilege = (privilege) => {
    const { myPrivileges } = this.state;
    console.log(myPrivileges);
    console.log(privilege);
  }

  handleAddPrivilegeClick = () => {
    const { privilegeValue, privileges, myPrivileges } = this.state;
    const { booker } = this.props;
    if (myPrivileges.some(p => p.id === privilegeValue)) {
      sweetAlert('Warning', 'You already have that privilege', 'warning');
      return;
    }
    api.addPrivilege([booker.user.username], privilegeValue)
      .then((r) => {
        if (r.status === 200) {
          const privilege = privileges.find(p => p.id === privilegeValue);
          myPrivileges.push(privilege);
          this.setState({ myPrivileges });
        }
      });
  }

  renderPrivilegeRow = (privilege, removePrivilegeAction) => (
    <List.Item key={privilege.id}>
      <List.Content floated="right">
        <Button onClick={removePrivilegeAction}>Remove</Button>
      </List.Content>
      <List.Content className="privilege-name">
        <Icon name="eye" />
        {privilege.name}
      </List.Content>
    </List.Item>
  )

  render() {
    const { show, booker, onClose } = this.props;
    const { dropdownOptions, myPrivileges } = this.state;
    console.log(booker);
    return (
      <Modal open={show} onClose={onClose} size="small" id="booker-modal">
        <Modal.Header>
          Booker details (
          {booker.user.username}
          )
        </Modal.Header>
        <Modal.Content>
          <Modal.Description>
            <h3>Fullname:</h3>
            {`${booker.user.last_name}, ${booker.user.first_name}`}
            <h3>Email</h3>
            {booker.user.email.length !== 0 ? booker.user.email : '[Empty]'}
            <h3>Privileges</h3>
            <div className="privilege-section">
              <List divided className="booker-list">
                {myPrivileges.map(
                  p => this.renderPrivilegeRow(p, this.handleRemovePrivilege),
                )}
              </List>
              <Dropdown
                placeholder="Privilege to add"
                selection
                onChange={this.handleDropdownChange}
                options={dropdownOptions}
              />
              <Button onClick={this.handleAddPrivilegeClick} color="blue">Add Privilege</Button>
            </div>
          </Modal.Description>
        </Modal.Content>
        <Modal.Actions>
          <Button onClick={onClose}>Close</Button>
        </Modal.Actions>
      </Modal>
    );
  }
}

BookerModal.propTypes = {
  show: PropTypes.bool.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  booker: PropTypes.object.isRequired,
  onClose: PropTypes.func.isRequired,
};

export default BookerModal;
