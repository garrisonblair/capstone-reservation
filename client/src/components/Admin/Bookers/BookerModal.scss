#booker-modal{
    .privilege-name{
        padding-top: 8px;
    }
    .privilege-section{
        border: 1px solid rgba(34,36,38,.15);
        padding: 8px;
    }
    .dropdown{
        margin-right: 4px;
    }
}
