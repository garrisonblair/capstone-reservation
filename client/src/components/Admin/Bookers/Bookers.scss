#bookers{
    .edit-button{
        display: block;
        margin: auto;
    }
}
.segment_container {
	display: grid;
    grid-template-columns: 1fr 1fr;
    .segment {
      margin: 0 2px;
    }
}
