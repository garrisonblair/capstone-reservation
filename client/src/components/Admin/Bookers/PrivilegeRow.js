import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, List } from 'semantic-ui-react';

class PrivilegeRow extends Component {
  handleRemovePrivilege = () => {

  }

  render() {
    const { privilege } = this.props;
    console.log(privilege);
    return (
      <List.Item>
        <List.Content floated="right">
          <Button>Remove</Button>
        </List.Content>
        <List.Content>
          {privilege.name}
        </List.Content>
      </List.Item>
    );
  }
}

PrivilegeRow.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  privilege: PropTypes.object.isRequired,
};

export default PrivilegeRow;
