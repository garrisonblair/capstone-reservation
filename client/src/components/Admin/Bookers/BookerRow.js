import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Table, Button, Icon } from 'semantic-ui-react';
import BookerModal from './BookerModal';


class BookerRow extends Component {
  state = {
    openModal: false,
  }

  onClickEditButton = () => {
    this.setState({ openModal: true });
  }

  closeModal = () => {
    const { syncBookers } = this.props;

    this.setState({ openModal: false });
    syncBookers();
  }

  render() {
    const { booker } = this.props;
    const { openModal } = this.state;
    return (
      <Table.Row key={booker.username}>
        <Table.Cell textAlign="center">
          {booker.username}
        </Table.Cell>
        <Table.Cell textAlign="center">
          {`${booker.first_name} ${booker.last_name}`}
        </Table.Cell>
        <Table.Cell textAlign="center">
          {booker.email}
        </Table.Cell>
        <Table.Cell textAlign="center">
          {booker.is_superuser ? <Icon name="check circle" color="green" /> : <Icon name="times circle" color="red" />}
        </Table.Cell>
        <Table.Cell textAlign="center">
          {booker.is_staff ? <Icon name="check circle" color="green" /> : <Icon name="times circle" color="red" />}
        </Table.Cell>
        <Table.Cell textAlign="center">
          {booker.is_active ? <Icon name="check circle" color="green" /> : <Icon name="times circle" color="red" />}
        </Table.Cell>
        <Table.Cell>
          <Button icon="edit" className="edit-button" onClick={this.onClickEditButton} />
        </Table.Cell>
        {openModal ? <BookerModal show booker={booker} onClose={this.closeModal} /> : null}
      </Table.Row>
    );
  }
}

BookerRow.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  booker: PropTypes.object.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  syncBookers: PropTypes.func.isRequired,
};
export default BookerRow;
