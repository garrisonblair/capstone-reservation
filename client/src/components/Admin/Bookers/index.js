import React, { Component } from 'react';
import { Table } from 'semantic-ui-react';
import './Bookers.scss';
import api from '../../../utils/api';
import BookerRow from './BookerRow';


class Bookers extends Component {
  state = {
    bookers: [],
  }

  componentDidMount() {
    this.syncBookers();
  }

  syncBookers = () => {
    api.getBookers()
      .then((r) => {
        if (r.status === 200) {
          this.setState({ bookers: r.data });
        }
      });
  }

  render() {
    const { bookers } = this.state;
    return (
      <div id="bookers">
        <h1>Bookers</h1>
        <Table>
          <Table.Header>
            <Table.Row key="-1">
              <Table.HeaderCell textAlign="center">
                Username
              </Table.HeaderCell>
              <Table.HeaderCell>
                {}
              </Table.HeaderCell>
            </Table.Row>
          </Table.Header>
          <Table.Body>
            {bookers.map(b => (
              <BookerRow booker={b} key={b.id} syncBookers={this.syncBookers} />
            ))}
          </Table.Body>
        </Table>
      </div>
    );
  }
}

export default Bookers;
