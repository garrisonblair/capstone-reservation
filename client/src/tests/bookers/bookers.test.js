/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import Bookers from '../../components/Admin/Bookers/index';

it('Shows bookers table', () => {
  const tree = renderer
    .create(
      <Bookers />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
