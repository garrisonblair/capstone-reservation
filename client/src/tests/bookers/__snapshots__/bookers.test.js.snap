// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows bookers table 1`] = `
<div
  id="bookers"
>
  <h1>
    Bookers
  </h1>
  <div
    className="ui loading segment"
  >
    <table
      className="ui table"
    >
      <thead
        className=""
      >
        <tr
          className=""
        >
          <th
            className="center aligned"
          >
            Username
          </th>
          <th
            className="center aligned"
          >
            Full name
          </th>
          <th
            className=""
          />
        </tr>
      </thead>
      <tbody
        className=""
      />
    </table>
  </div>
</div>
`;
